### PR TITLE
HPCC-13889 Add support for graphs in Cassandra WU plugin

### PR DESCRIPTION
**Pull request by richardkchapman**
Cleanup unused functions and code in workunit API.

Add support for creating and iterating graphs.

Note: graph progress is NOT supported yet.

Signed-off-by: Richard Chapman rchapman@hpccsystems.com

### CODE REVIEW

**Comment by richardkchapman** [comment]
@ghalliday @jakesmith please review

File: common/workunit/workunit.cpp [hunk 1]
```diff
@@ -6390,27 +6293,24 @@ IPropertyTree *CLocalWorkUnit::getUnpackedTree(bool includeProgress) const
...
-            }
-        }
+        IPropertyTree &graph = iter->query();
+        graphs.append(*new CLocalWUGraph(*this, LINK(&graph)));
+    }
+}
+
+void CLocalWorkUnit::loadGraphs(bool heavy) const
+{
+    if (graphsCached < (heavy ? 2 : 1))
```
Comment (jakesmith) [inline common/workunit/workunit.cpp]: could do with a comment here, to clarify how heavy interacts with graphCached.

File: plugins/cassandra/cassandrawu.cpp [hunk 2]
```diff
@@ -3380,7 +3506,20 @@ class CCasssandraWorkUnitFactory : public CWorkUnitFactory, implements ICassandr
...
         return executeQuery(session, select);
+    }
 
+    const CassResult *fetchDataForWuidAndKey(const CassandraXmlMapping *mappings, const char *wuid, const char *key) const
+    {
+        assertex(wuid && *wuid);
+        StringBuffer names;
+        StringBuffer tableName;
+        getFieldNames(mappings+2, names, tableName);  // mappings+3 means we don't return the partition or wuid columns. We do return the key.
+        VStringBuffer selectQuery("select %s from %s where partition=? and wuid=? and %s=?;", names.str()+1, tableName.str(), mappings[2].columnName);
```
Comment (jakesmith) [inline plugins/cassandra/cassandrawu.cpp]: I'm curious, why names.str()+1 ?

**Comment by jakesmith** [comment]
Very minor comments, it looks clean to me.
@ghalliday please take a 2nd look.

File: plugins/cassandra/cassandrawu.cpp (hunk 2 above)
Comment (richardkchapman) [inline plugins/cassandra/cassandrawu.cpp]: names returns a leading comma.

File: ecl/wutest/wutest.cpp [hunk 3]
```diff
@@ -755,6 +745,123 @@ class WuTest : public CppUnit::TestFixture
...
+
+        // Light then heavy from a single wu
+        wu.setown(factory->openWorkUnit(wuid));
+        ASSERT(streq(wu->queryWuid(), wuid));
+        it.setown(&wu->getGraphsMeta(GraphTypeActivities));
+        numIterated = 0;
+        ForEach (*it)
+        {
+            ASSERT(it->query().getType() == GraphTypeActivities);
+            ASSERT(streq(graph->getLabel(s).str(),"graphLabel"));
```
Comment (ghalliday) [inline ecl/wutest/wutest.cpp]: it->query().getLabel()?

File: ecl/wutest/wutest.cpp [hunk 4]
```diff
@@ -755,6 +745,123 @@ class WuTest : public CppUnit::TestFixture
...
+
+        // then the heavy meta
+        wu.setown(factory->openWorkUnit(wuid));
+        ASSERT(streq(wu->queryWuid(), wuid));
+        Owned<IConstWUGraphIterator> it2 = &wu->getGraphs(GraphTypeActivities);
+        numIterated = 0;
+        ForEach (*it2)
+        {
+            ASSERT(it2->query().getType() == GraphTypeActivities);
+            ASSERT(streq(it2->query().getLabel(s).str(),"graphLabel"));
```
Comment (ghalliday) [inline ecl/wutest/wutest.cpp]: Is this really testing the heavy meta?

File: ecl/wutest/wutest.cpp [hunk 5]
```diff
@@ -755,6 +745,123 @@ class WuTest : public CppUnit::TestFixture
...
+            numIterated++;
+        }
+        ASSERT(numIterated==2);
+
+        it.setown(&wu->getGraphsMeta(GraphTypeActivities));
+        numIterated = 0;
+        ForEach (*it)
+        {
+            ASSERT(it->query().getType() == GraphTypeActivities);
+            ASSERT(streq(graph->getLabel(s).str(),"graphLabel"));
```
Comment (ghalliday) [inline ecl/wutest/wutest.cpp]: it->query()

File: plugins/cassandra/cassandrawu.cpp [hunk 6]
```diff
@@ -2203,6 +2276,8 @@ class CCassandraWorkUnit : public CPersistedWorkUnit
                 // empty newly-created WU, it is unnecessary.
                 //deleteChildren(wuid);
 
+                // MORE can use the table
```
Comment (ghalliday) [inline plugins/cassandra/cassandrawu.cpp]: Is this comment still relevant?

File: plugins/cassandra/cassandrawu.cpp [hunk 7]
```diff
@@ -3380,7 +3506,20 @@ class CCasssandraWorkUnitFactory : public CWorkUnitFactory, implements ICassandr
...
         select.bindString(1, wuid);
         return executeQuery(session, select);
+    }
 
+    const CassResult *fetchDataForWuidAndKey(const CassandraXmlMapping *mappings, const char *wuid, const char *key) const
+    {
+        assertex(wuid && *wuid);
+        StringBuffer names;
+        StringBuffer tableName;
+        getFieldNames(mappings+2, names, tableName);  // mappings+3 means we don't return the partition or wuid columns. We do return the key.
```
Comment (ghalliday) [inline plugins/cassandra/cassandrawu.cpp]: comment refers to +3

**Comment by ghalliday** [comment]
@richardkchapman some very small comments

File: ecl/wutest/wutest.cpp (hunk 3 above)
Comment (richardkchapman) [inline ecl/wutest/wutest.cpp]: Fixed

File: ecl/wutest/wutest.cpp (hunk 4 above)
Comment (richardkchapman) [inline ecl/wutest/wutest.cpp]: It is, though you'd have to know the internal implementations to know it. I did check the tracing to be sure it fetched the full information in this case.

File: ecl/wutest/wutest.cpp (hunk 5 above)
Comment (richardkchapman) [inline ecl/wutest/wutest.cpp]: Fixed

File: plugins/cassandra/cassandrawu.cpp (hunk 6 above)
Comment (richardkchapman) [inline plugins/cassandra/cassandrawu.cpp]: Yes. I should really consider using the list of all child tables here.

File: plugins/cassandra/cassandrawu.cpp (hunk 7 above)
Comment (richardkchapman) [inline plugins/cassandra/cassandrawu.cpp]: Fixed

**Pull request merged**